File: maximal-square.nim
type Arr[T] = ptr UncheckedArray[T]

proc maximalSquare(a: Arr[Arr[char]], m: int, ns: Arr[cint]): cint {.exportc.} =
  if m == 0: return 0
  let n = ns[0].int
  var
    s = newSeq[int](n)
    l = 0
  for i in 0..<m:
    var ul = 0
    for j in 0..<n:
      let t = ul
      ul = s[j]
      s[j] = if a[i][j] == '1': min(min(t, if j>0: s[j-1] else: 0), s[j])+1 else: 0
      l = max(l, s[j])
  cast[cint](l*l)

File: maximal-square.cc
// Maximal Square
#define FOR(i, a, b) for (int i = (a); i < (b); i++)
#define REP(i, n) for (int i = 0; i < (n); i++)

class Solution {
public:
  int maximalSquare(vector<vector<char>> &a) {
    if (a.empty()) return 0;
    int m = a.size(), n = a[0].size(), ans = 0;
    vector<int> s(n, 0);
    REP(i, m) {
      int ul = 0;
      REP(j, n) {
        int t = ul;
        ul = s[j];
        s[j] = a[i][j] == '1' ? min(min(t, j?s[j-1]:0), s[j])+1 : 0;
        ans = max(ans, s[j]);
      }
    }
    return ans*ans;
  }
};
